src/type/t_visit.ml

```diff
@@ -0,0 +1,803 @@
+            (a, FcnApp (f, es) @@ oe, ty04)
+        | Some (ty03s) when List.length ty03s = n+1 && n > 1 ->
+            let ty04s, ty05 =
+              match List.rev ty03s with
+              | ty05 :: r_ty04s -> List.rev r_ty04s, ty05
+              | _ -> error ~at:oe "Bad type annotation"
+            in
+            check0 (TFun (TPrd ty04s, ty05)) ty01;
+            iter3 (fun e -> check0 ~at:e) es ty04s ty02s;
+            (a, FcnApp (f, es) @@ oe, ty05)
+        | _ ->
+            error ~at:oe "Bad type annotation"
+        end
+    | Arrow (e, f) ->
+        let a, e, ty01 = self#expr scx a e in
+        let a, f, ty02 = self#expr scx a f in
+        begin match query oe Props.tpars_prop with
+        | None ->
+            check0 (TAtm TAIdv) ty01;
+            check0 (TAtm TAIdv) ty02;
+            (a, Arrow (e, f) @@ oe, TAtm TAIdv)
+        | Some ([ ty03 ; ty04 ]) ->
+            check0 (TSet ty03) ty01;
+            check0 (TSet ty04) ty02;
+            (a, Arrow (e, f) @@ oe, TSet (TFun (ty03, ty04)))
+        | _ ->
+            error ~at:oe "Bad type annotation"
+        end
+    | Product es ->
+        let a, es, ty01s =
+          List.fold_left begin fun (a, r_es, r_ty0s) e ->
+            let a, e, ty0 = self#expr scx a e in
+            (a, e :: r_es, ty0 :: r_ty0s)
+          end (a, [], []) es |>
+          fun (a, r_es, r_ty0s) -> (a, List.rev r_es, List.rev r_ty0s)
+        in
+        begin match query oe Props.tpars_prop with
+        | None ->
+            List.iter2 (fun e -> check0 ~at:e (TAtm TAIdv)) es ty01s;
+            (a, Product es @@ oe, TAtm TAIdv)
+        | Some ty02s ->
+            iter3 (fun e -> check0 ~at:e) es ty02s ty01s;
+            (a, Product es @@ oe, TSet (TPrd ty02s))
+        end
+    | Tuple es ->
+        let a, es, ty01s =
+          List.fold_left begin fun (a, r_es, r_ty0s) e ->
+            let a, e, ty0 = self#expr scx a e in
+            (a, e :: r_es, ty0 :: r_ty0s)
+          end (a, [], []) es |>
+          fun (a, r_es, r_ty0s) -> (a, List.rev r_es, List.rev r_ty0s)
+        in
+        begin match query oe Props.tpars_prop with
+        | None ->
+            List.iter2 (fun e -> check0 ~at:e (TAtm TAIdv)) es ty01s;
+            (a, Product es @@ oe, TAtm TAIdv)
+        | Some ty02s ->
+            iter3 (fun e -> check0 ~at:e) es ty02s ty01s;
+            (a, Product es @@ oe, TPrd ty02s)
+        end
+    (* FIXME *)
+    (*
+    | Rect fs ->
+        let a, fs = List.fold_left begin fun (a, fs) (s, e) ->
+          let a, e = self#expr scx a e in
+          (a, (s, e) :: fs)
+        end (a, []) fs in
+        let fs = List.rev fs in
+        (a, Rect fs @@ oe)
+    | Record fs ->
+        let a, fs = List.fold_left begin fun (a, fs) (s, e) ->
+          let a, e = self#expr scx a e in
+          (a, (s, e) :: fs)
+        end (a, []) fs in
+        let fs = List.rev fs in
+        (a, Record fs @@ oe)
+    *)
```

Comment:
I removed it.  There is a lot of code around the type system that I wrote without knowing what the type system would actually be or how it would work...